### PR TITLE
paho-mqtt-c: add version 1.3.12, support conan v2

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **paho-mqtt-c/***

- add version 1.3.12
- support conan v2
- sort data by version number in config.yml/conandata.yml
- add patch descriptions

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by lbakman — APPROVED** [review]
I have checked pull request using conan 1.58.0.
This looks good to me.

**Comment by toge** [comment]
@lbakman 
Thank you for your feedback!
CI errors are caused by my mistake on MSVC build.
It is fixed probably.

**Review by lbakman — APPROVED** [review]
Built and tested on Linux, using conan-center hooks and 'export CONAN_HOOK_ERROR_LEVEL=40'.
Looks good to me.

File: recipes/paho-mqtt-c/all/conanfile.py [hunk 1]
```diff
@@ -52,80 +49,89 @@ def config_options(self):
...
+        deps.generate()
 
     def _patch_source(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
-                              "SET(CMAKE_MODULE_PATH \"${CMAKE_SOURCE_DIR}/cmake/modules\")",
-                              "LIST(APPEND CMAKE_MODULE_PATH \"${CMAKE_SOURCE_DIR}/cmake/modules\")")
+        apply_conandata_patches(self)
+        replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"),
```
Comment (prince-chrismc) [inline recipes/paho-mqtt-c/all/conanfile.py:100]: Seems like it's needed with the new stuff https://github.com/conan-io/conan/blob/e0a8ee058bc8dc2d9811b8aeb6999f69aeb78d85/conan/tools/cmake/toolchain/blocks.py#L450

File: recipes/paho-mqtt-c/all/test_package/conanfile.py [hunk 2]
```diff
@@ -1,20 +1,32 @@
-from conans import ConanFile, CMake, tools
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
```
Comment (prince-chrismc) [inline recipes/paho-mqtt-c/all/test_package/conanfile.py:3]:
```suggestion
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
```

File: recipes/paho-mqtt-c/all/test_package/conanfile.py [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
+from conan.tools.env import VirtualBuildEnv
```
Comment (prince-chrismc) [inline recipes/paho-mqtt-c/all/test_package/conanfile.py:4]:
```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
Bump of very minor comments, just some clean up

Excellent work!

File: recipes/paho-mqtt-c/all/test_package/conanfile.py (hunk 2 above)
Comment (toge) [reply recipes/paho-mqtt-c/all/test_package/conanfile.py:3]: remove unused modules

File: recipes/paho-mqtt-c/all/conanfile.py (hunk 1 above)
Comment (toge) [reply recipes/paho-mqtt-c/all/conanfile.py:94]: @prince-chrismc 
Thanks you for lots of comments.

Last thing about CMAKE_MODULE_PATH, I think there are no way with `find_paths` to solve the issue original CMakeLists.txt sets CMAKE_MODULE PATH with single value.
Any advice would be appreciated.

**Review by uilianries — APPROVED** [review]
LGTM

**Review by lbakman — APPROVED** [review]
LGTM

**Review by jwillikers — APPROVED** [review]

**Pull request merged**